example/poisson_gmg/parthenon_app_inputs.cpp

```diff
@@ -0,0 +1,110 @@
+//========================================================================================
+// (C) (or copyright) 2023. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#include <math.h>
+#include <sstream>
+#include <string>
+
+#include <parthenon/package.hpp>
+
+#include "config.hpp"
+#include "defs.hpp"
+#include "poisson_package.hpp"
+#include "utils/error_checking.hpp"
+
+using namespace parthenon::package::prelude;
+using namespace parthenon;
+
+// *************************************************//
+// redefine some weakly linked parthenon functions *//
+// *************************************************//
+
+namespace poisson_example {
+
+void ProblemGenerator(Mesh *pm, ParameterInput *pin, MeshData<Real> *md) {
+  auto pmb = md->GetBlockData(0)->GetBlockPointer();
+  const int ndim = md->GetMeshPointer()->ndim;
+
+  Real x0 = pin->GetOrAddReal("poisson", "x0", 0.0);
+  Real y0 = pin->GetOrAddReal("poisson", "y0", 0.0);
+  Real z0 = pin->GetOrAddReal("poisson", "z0", 0.0);
+  Real radius0 = pin->GetOrAddReal("poisson", "radius", 0.1);
+  Real interior_D = pin->GetOrAddReal("poisson", "interior_D", 1.0);
+  Real exterior_D = pin->GetOrAddReal("poisson", "exterior_D", 1.0);
+
+  auto desc =
+      parthenon::MakePackDescriptor<poisson_package::rhs, poisson_package::res_err,
+                                    poisson_package::u, poisson_package::D,
+                                    poisson_package::exact>(md);
+  auto pack = desc.GetPack(md);
+
+  constexpr auto te = poisson_package::te;
+  using TE = parthenon::TopologicalElement;
+  auto &cellbounds = pmb->cellbounds;
+  auto ib = cellbounds.GetBoundsI(IndexDomain::entire, te);
+  auto jb = cellbounds.GetBoundsJ(IndexDomain::entire, te);
+  auto kb = cellbounds.GetBoundsK(IndexDomain::entire, te);
+  pmb->par_for(
+      "Poisson::ProblemGenerator", 0, pack.GetNBlocks() - 1, kb.s, kb.e, jb.s, jb.e, ib.s,
```

Comment:
It only occurred to me now, but the `pmb->par_for` is kind of a leftover from Athena++ block approach.
It might be worth to add a `pack->par_for` that includes the iteration over blocks at some point.

Reply:
:+1: Alternatively we just use `parthenon::par_for(DEFAULT_LOOP_PATTERN` or have an overload that just automatically uses the default pattern.

If we did `pack->par_for` though, we could probably have an overload that automatically chose the right bounds for some cases...

Reply:
I probably used `pmb->par_for` because I copied and pasted something else. I would probably lean toward using free functions because they are simpler and less interdependent (i.e. `parthenon::par_for`). I agree it would be nice to have an overload that just uses the default loop pattern there.

I think that there are too many possible cases for packs to choose the correct loop bounds, especially now that different blocks in a pack can have different sizes.

Reply:
Is there any downstream use where you have a loop over a pack that does *not* iterate over all blocks?
As far as I can tell, in AthenaPK all loops over a pack always go over all blocks so for us a `pack->par_for` that provides a default index to go over the blocks would be a win (in terms of less boilerplate).

Reply:
I think it is true that probably every kernel we write goes over all blocks of the pack as the first index. I guess we could add a loop interface to the pack that automatically adds this first index. I am not sure how much boilerplate this would remove though, since you would have to still get the (k, j, i) etc. ranges by hand.

Reply:
Is there a reason why we have to get the (k,j,i) ranges by hand?

Reply:
It is now possible that not all blocks in a pack are the same size, so you can't generically assume a fixed (k,j,i) index range. That being said, it would certainly be possible to write an overload that assumes the block size in a pack is fixed (which is true almost all of the time) and automatically pulls out the k,j,i index range as well. I would advocate against making these member functions though and rather have them be something like 
```c++
parthenon::par_for_bkji(DEFAULT_LOOP_PATTERN, pack, [](int b, int k, int j, int i){});
```

Reply:
Ah, ok. At the moment, is it the case that only the multigrid code uses packs that don't have fixed-size blocks?

Reply:
yeah, right now it is only MG that can end up with variable sized blocks.